internal/abci/abci.go

```diff
@@ -66,21 +64,30 @@ func NewAbciApp(ctx context.Context, cfg *AbciConfig, snapshotter SnapshotModule
 		log: log,
 
 		validatorAddressToPubKey: make(map[string][]byte),
-		txCache:                  make(map[string]bool),
+		verifiedTxns:             make(map[chainHash]struct{}),
 	}
+	app.forks.FromMap(cfg.ForkHeights)
 
 	tx, err := db.BeginOuterTx(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to begin outer tx: %w", err)
 	}
 	defer tx.Rollback(ctx)
 
-	app.forks.FromMap(cfg.ForkHeights)
+	// Populate the validatorAddressToPubKey field.
+	validators, err := txRouter.GetValidators(ctx, tx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get validators: %w", err)
+	}
+	for _, val := range validators {
+		addr, err := pubkeyToAddr(val.PubKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert pubkey to address: %w", err)
+		}
+
+		app.validatorAddressToPubKey[addr] = val.PubKey
```

Comment:
Also somewhat unrelated other than being in abci, but this was moved from the `Info` method, which is not part of cometbft's consensus connection methods, making it not thread safe.  The map isn't guarded, so it should only be touched in constructor or the consensus connection methods that are used sequentially.

Honestly I don't know why this was done in `Info`, but my guess is that we were using it like a constructor where we only had DB there.  But now we've got DB from the outset, so this is an easy fix now.